controllers/kubeadmconfig_controller.go

```diff
@@ -140,6 +140,23 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, err
 	}
 
+	// If we've already embedded a time-limited join token into a config, but are still waiting for the token to be used, refresh it
+	if config.Status.Ready {
```

Comment:
I wonder if this is an appropriate time for a switch? It can be out of scope for this PR.

Something like this:
```
switch {
  case machine.Spec.Bootstrap.Data != nil && !config.Status.Ready:
  case config.Status.Ready && machine.Status.InfrastructureReady:
  case config.Status.Ready && !machine.Status.InfrastructureReady:
  default:
}
```

I find it easier to see edge cases and make sure every path is handled explicitly